cpp/include/raft/error.hpp

```diff
@@ -0,0 +1,329 @@
+  throw raft::nccl_error(
+    std::string{"NCCL error encountered at: " + std::string{file} + ":" +
+                std::to_string(line) + ": " + std::to_string(error) + " " +
+                ncclGetErrorString(error)});
+}
+
+#define _CURAND_ERR_TO_STR(err) \
+  case err:                     \
+    return #err;
+inline auto curand_error_to_string(curandStatus_t err) -> const char* {
+  switch (err) {
+    _CURAND_ERR_TO_STR(CURAND_STATUS_SUCCESS);
+    _CURAND_ERR_TO_STR(CURAND_STATUS_VERSION_MISMATCH);
+    _CURAND_ERR_TO_STR(CURAND_STATUS_NOT_INITIALIZED);
+    _CURAND_ERR_TO_STR(CURAND_STATUS_ALLOCATION_FAILED);
+    _CURAND_ERR_TO_STR(CURAND_STATUS_TYPE_ERROR);
+    _CURAND_ERR_TO_STR(CURAND_STATUS_OUT_OF_RANGE);
+    _CURAND_ERR_TO_STR(CURAND_STATUS_LENGTH_NOT_MULTIPLE);
+    _CURAND_ERR_TO_STR(CURAND_STATUS_DOUBLE_PRECISION_REQUIRED);
+    _CURAND_ERR_TO_STR(CURAND_STATUS_LAUNCH_FAILURE);
+    _CURAND_ERR_TO_STR(CURAND_STATUS_PREEXISTING_FAILURE);
+    _CURAND_ERR_TO_STR(CURAND_STATUS_INITIALIZATION_FAILED);
+    _CURAND_ERR_TO_STR(CURAND_STATUS_ARCH_MISMATCH);
+    _CURAND_ERR_TO_STR(CURAND_STATUS_INTERNAL_ERROR);
+    default:
+      return "CURAND_STATUS_UNKNOWN";
+  };
+}
+#undef _CURAND_ERR_TO_STR
+
+inline void throw_curand_error(curandStatus_t error, const char* file,
+                               unsigned int line) {
+  throw raft::curand_error(
+    std::string{"cuRAND error encountered at: " + std::string{file} + ":" +
+                std::to_string(line) + ": " + std::to_string(error) + " " +
+                curand_error_to_string(error)});
+}
+
+// FIXME: unnecessary once CUDA 10.1+ becomes the minimum supported version
+#define _CUSPARSE_ERR_TO_STR(err) \
+  case err:                       \
+    return #err;
+inline auto cusparse_error_to_string(cusparseStatus_t err) -> const char* {
+#if defined(CUDART_VERSION) && CUDART_VERSION >= 10100
+  return cusparseGetErrorString(status);
+#else   // CUDART_VERSION
+  switch (err) {
+    _CUSPARSE_ERR_TO_STR(CUSPARSE_STATUS_SUCCESS);
+    _CUSPARSE_ERR_TO_STR(CUSPARSE_STATUS_NOT_INITIALIZED);
+    _CUSPARSE_ERR_TO_STR(CUSPARSE_STATUS_ALLOC_FAILED);
+    _CUSPARSE_ERR_TO_STR(CUSPARSE_STATUS_INVALID_VALUE);
+    _CUSPARSE_ERR_TO_STR(CUSPARSE_STATUS_ARCH_MISMATCH);
+    _CUSPARSE_ERR_TO_STR(CUSPARSE_STATUS_EXECUTION_FAILED);
+    _CUSPARSE_ERR_TO_STR(CUSPARSE_STATUS_INTERNAL_ERROR);
+    _CUSPARSE_ERR_TO_STR(CUSPARSE_STATUS_MATRIX_TYPE_NOT_SUPPORTED);
+    default:
+      return "CUSPARSE_STATUS_UNKNOWN";
+  };
+#endif  // CUDART_VERSION
+}
+#undef _CUSPARSE_ERR_TO_STR
+
+inline void throw_cusparse_error(cusparseStatus_t error, const char* file,
+                                 unsigned int line) {
+  throw raft::cusparse_error(
+    std::string{"cuSparse error encountered at: " + std::string{file} + ":" +
+                std::to_string(line) + ": " + std::to_string(error) + " " +
+                cusparse_error_to_string(error)});
+}
```

Comment:
We have this check already defined in https://github.com/rapidsai/raft/blob/branch-0.15/cpp/include/raft/linalg/cusolver_wrappers.h, better to update the existing code there?

Reply:
> It makes sense to put the basic assertions in here. But putting assertions related to curand/cusparse/nccl in the same header file means that we'll introduce dependencies on these even if the including source doesn't use these components. I'd highly prefer if we could separate these out.

Yeah... so the pros is to keep the related code in one place so it gets easy to enforce consistency in error handling, but the problem you mentioned is also valid (especially NCCL can be problematic if we just compile for single-GPU).

I'm fine with either approach as long as we are consistent and can maintain consistency.

I will move

CUDA related ones to https://github.com/rapidsai/raft/blob/branch-0.15/cpp/include/raft/cudart_utils.h

cuSparse related ones to https://github.com/rapidsai/raft/blob/branch-0.15/cpp/include/raft/sparse/cusparse_wrappers.h

NCCL related ones to https://github.com/rapidsai/raft/blob/branch-0.15/cpp/include/raft/comms/std_comms.hpp

I should addd CUBLAS_TRY and CUSOLVER_TRY to https://github.com/rapidsai/raft/blob/branch-0.15/cpp/include/raft/linalg/cublas_wrappers.h and https://github.com/rapidsai/raft/blob/branch-0.15/cpp/include/raft/linalg/cusolver_wrappers.h

I may drop CURAND related ones for now, but they should be added back when we have curand_wrappers.hpp

And are you guys using XXX_CHECK in RAFT from cuML? If I replace them with XXX_TRY, will this break cuML?

And any thoughts about XXX_TRY vs XXX_CHECK? cuDF is using XXX_TRY, so I am just following cuDF convention but this is undesirable for cuML, I am open to discussion.

Reply:
+1 for CUBLAS_TRY and CUSOLVER_TRY they will be needed in #12 

I'm fine with XXX_TRY for consistency with cuDF. It also captures the underlying exception mechanism a bit better in the name.

Reply:
@seunghwak sorry for the delayed response.

Yes, we are using the *_CHECK macros everywhere inside cuML. So, replacing those will certainly break our codebase. Moreover, the C-style '%' modifiers provides more readable code than the '<<' style C++ syntax (purely my opinion).

Reply:
OK, better keep both *_TRY and *_CHECK in the meantime (better pick one eventually).

There can be a long debate in printf-ish vs cout-ish (e.g. https://stackoverflow.com/questions/2872543/printf-vs-cout-in-c).

My biggest concern was the type safety (in the above link) but it seems like most host compilers generates a warning for this and we are OK unless we ignore warnings. This wasn't a case with nvcc and leads to a weird runtime behavior but this is irrelevant here as this is a host side error checking mechanism. We may better keep the C-ish style (and add C++ style in the future if necessary).